Use an EraMap to specify rules for the era.

nc6 opened the pull request:
The way we implement the era-parametrised rules leads to a somewhat
annoying situation. We have two choices when creating an STS "instance".
We can either declare it's specific to a given era:

```
instance STS UTXO (ShelleyEra c)
```

Or we can declare it's valid for _all_ eras:

```
instance STS UTXO era
```

Of course, what we'd really like is to say "this instance should be
valid for all eras going forth, until it's overridden". But this is a
little tricky to do at the type level, and we'd have to introduce some
dependently-typed wizardry to manage it.

This problem shows itself particularly with the update system, for
example. The update system is not modified in Allegra, Mary or Alonzo.
However, to support overriding it in a future era (which we need), we
would have to move from the "all era" style to the "current era only"
style. So the Shelley declaration would become valid for Shelley only,
and we'd have to introduce (identical) instances in Allegra, Mary and
Alonzo. This would both be substantial duplication and would muddy the
waters of what's actually overridden in each era.

This PR presents an alternative, "lightweight" solution to this problem,
based upon Damian's work in #2041. We add an additional layer of
indirection in the form of a type family which maps symbols (e.g.
"UTXO") into a specific implementation of the rule. We then have a map
of these per era (the `EraMap`). In future eras, we provide exactly this
same map, which identifies which rules come from which era. Obviously,
constraints guide which rules can be connected to each other.

This approach also removes the need for orphan STS instances in future
eras, since we would now define a "new" UTXO rule in a future era, for
example.

nc6 committed:
Use an EraMap to specify rules for the era.

The way we implement the era-parametrised rules leads to a somewhat
annoying situation. We have two choices when creating an STS "instance".
We can either declare it's specific to a given era:

```
instance STS UTXO (ShelleyEra c)
```

Or we can declare it's valid for _all_ eras:

```
instance STS UTXO era
```

Of course, what we'd really like is to say "this instance should be
valid for all eras going forth, until it's overridden". But this is a
little tricky to do at the type level, and we'd have to introduce some
dependently-typed wizardry to manage it.

This problem shows itself particularly with the update system, for
example. The update system is not modified in Allegra, Mary or Alonzo.
However, to support overriding it in a future era (which we need), we
would have to move from the "all era" style to the "current era only"
style. So the Shelley declaration would become valid for Shelley only,
and we'd have to introduce (identical) instances in Allegra, Mary and
Alonzo. This would both be substantial duplication and would muddy the
waters of what's actually overridden in each era.

This PR presents an alternative, "lightweight" solution to this problem,
based upon Damian's work in #2041. We add an additional layer of
indirection in the form of a type family which maps symbols (e.g.
"UTXO") into a specific implementation of the rule. We then have a map
of these per era (the `EraMap`). In future eras, we provide exactly this
same map, which identifies which rules come from which era. Obviously,
constraints guide which rules can be connected to each other.

This approach also removes the need for orphan STS instances in future
eras, since we would now define a "new" UTXO rule in a future era, for
example.

nc6 commented:
Note this is a draft at the moment since I haven't made all the changes in future eras. But this should demonstrate the approach, enough to solicit views :-)

mrBliss commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/EraMap.hs:
```diff
@@ -0,0 +1,77 @@
+
+-- | This module provides the mapping from conceptual rule names to concrete
+-- rules for the Shelley era.
+module Shelley.Spec.Ledger.STS.EraMap () where
```
So this module will only contain the Shelley mapping. How should the modules containing the other mappings be called? `Cardano.Ledger.ShelleyMA.STS.EraMap` and/or `Cardano.Ledger.Allegra/Mary.STS.EraMap`, etc?

For consistency, what about calling this module `Cardano.Ledger.Shelley.STS.EraMap`?

mrBliss replied:
Minor: when I see the `-Map` in `EraMap` I think of `Map` the data type instead of the mapping. What about `EraMapping`?

nc6 replied:
At some point I want to move the whole `Shelley` codebase to `Cardano.Ledger.Shelley`. And `STS` to `Rule`, indeed...

dnadales replied:
Sorry I made a similar comment about this. So please ignore it.

mrBliss commented on shelley/chain-and-ledger/executable-spec/src/Cardano/Ledger/Core.hs:
```diff
@@ -55,3 +61,6 @@ type SerialisableData t = (FromCBOR t, ToCBOR t)
 type AnnotatedData t = (FromCBOR (Annotator t), ToCBOR t)
+
+-- | Era STS map
+type family EraSTS (k :: Symbol) era :: Type
```
I like the `Era-` prefix, we should use it for all of these type families :smile: 

Just a thought: what about `EraRule`?

nc6 replied:
Yep, that's nicer :-)

dnadales replied:
Can't we remove the `Era` part. In this way we avoid the repetition of eras: for instance `Core.EraRule "POOLREAP" era`. Why not simply `Core.Rule "POOLREAP" era`.

mrBliss replied:
I would drop the `Core.` prefix instead. I'm in favour of using the `Era-` prefix for all such type families so that we no longer have name conflicts such as `Core.Crypto` vs the `Crypto` class, `Core.TxBody` vs the various concrete `TxBody` types, ...

mrBliss commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/EraMap.hs:
```diff
@@ -0,0 +1,77 @@
+import Shelley.Spec.Ledger.STS.Utxo (UTXO)
+import Shelley.Spec.Ledger.STS.Utxow (UTXOW)
+
+type instance Core.EraSTS "BBODY" (ShelleyEra c) = BBODY (ShelleyEra c)
```
I would be nice if this were more inheritable/reusable, e.g., "use the mapping from this era, but override only these rules".

nc6 replied:
Yes, definitely. And here one moves towards type-level maps etc. Is it worth it? Maybe I'll investigate...

dnadales replied:
Yup, I was thinking about the same.

mrBliss reviewed (commented):
Good idea!

Thinking out loud: this mapping solves the problem at the cost of some extra boilerplate. When I think about solutions to the extra boilerplate, I find myself reaching for the type class machinery again, which is exactly where we are at before this PR.

To avoid identical instances, could we use deriving via? For example, the  TICKF rule instance of Allegra can be derived via Shelley's instance? I'm not sure if that's even possible with the data and type families involved.

nc6 commented:
> To avoid identical instances, could we use deriving via? For example, the TICKF rule instance of Allegra can be derived via Shelley's instance? I'm not sure if that's even possible with the data and type families involved.

I looked at that a while ago, and came to the conclusion of basically no (or, not without being significantly cleverer than I am!) :-( The problem tends to be that the various type/data families may or may not be themselves era parametrised - if so, they need to be updated, but there's nothing type-wise saying that they must even take that parameter. TH could do it, but I don't think `DerivingVia` can.

JaredCorduan reviewed (commented):
I'm for it, thanks!

dnadales commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/EraMapping.hs:
```diff
@@ -0,0 +1,77 @@
+
+-- | This module provides the mapping from conceptual rule names to concrete
+-- rules for the Shelley era.
+module Shelley.Spec.Ledger.STS.EraMapping () where
```
Minor: there's an inconsistency regarding the places where we define these instances. In ShelleyMA they are defined in: `module Cardano.Ledger.ShelleyMA.Rules.EraMapping`.

dnadales commented:
The description of this PR mentions that EraMap constrains which rules can be connected to each other, but does it? And if so how? I think it was the rules itself that defined these constraints.

dnadales commented:
Thinking about developers ergonomics, how do we know the rules we have to define per-era? Look at what's defined in the previous ones? Or compile, get the missing instances, and add them?

mrBliss commented:
> Thinking about developers ergonomics, how do we know the rules we have to define per-era? Look at what's defined in the previous ones? Or compile, get the missing instances, and add them?

Good point. Errors about missing type family instances are rather obscure, IIRC, as the type family application just doesn't reduce. Missing type class instances give much clearer error messages.

dnadales reviewed (commented):
First of all thank you for the PR description! It really helped me to ask myself some questions before starting the review, and find the answers while reviewing :)

I like this approach even if we cannot solve the boilerplate problem, and I think it's an improvement w.r.t. the current situation.

nc6 committed:
Drop injectivity requirement for PredicateFailure

This is split apart from #2060, since it's logically distinct. The
motivation is that we will need this to have multiple UTXO rules in
different eras sharing the same failure.

nc6 committed:
Use an EraMap to specify rules for the era

The way we implement the era-parametrised rules leads to a somewhat
annoying situation. We have two choices when creating an STS "instance".
We can either declare it's specific to a given era:

instance STS UTXO (ShelleyEra c)
Or we can declare it's valid for all eras:

instance STS UTXO era
Of course, what we'd really like is to say "this instance should be
valid for all eras going forth, until it's overridden". But this is a
little tricky to do at the type level, and we'd have to introduce some
dependently-typed wizardry to manage it.

This problem shows itself particularly with the update system, for
example. The update system is not modified in Allegra, Mary or Alonzo.
However, to support overriding it in a future era (which we need), we
would have to move from the "all era" style to the "current era only"
style. So the Shelley declaration would become valid for Shelley only,
and we'd have to introduce (identical) instances in Allegra, Mary and
Alonzo. This would both be substantial duplication and would muddy the
waters of what's actually overridden in each era.

This PR presents an alternative, "lightweight" solution to this problem,
based upon Damian's work in #2041. We add an additional layer of
indirection in the form of a type family which maps symbols (e.g.
"UTXO") into a specific implementation of the rule. We then have a map
of these per era (the EraMap). In future eras, we provide exactly this
same map, which identifies which rules come from which era. Obviously,
constraints guide which rules can be connected to each other.

This approach also removes the need for orphan STS instances in future
eras, since we would now define a "new" UTXO rule in a future era, for
example.

nc6 commented:
Closing in favour of #2080
